### PR TITLE
Refactor logger 

### LINKED ISSUE #18: Feature additions for the Logger
The current state of the logger is acceptable for use across the current code base, but is in general quite simple and lacks much control in how a user may want to use it. As such below is a list of features we should aim to add to the logger to improve its functionality.

- [ ] **Add the ability to control the verbosity of the logger.** The current version is quite loud in it tells you a lot of information about where the logging occurred. This can be quite helpful when we may want verbose logging, but should not be the default mode. We should add the ability to turn on and off verbose logging.
- [ ] **Enable the ability to set if we output logging data to a file or console or both.** More control over the ability to set what stream we want to log to would be super useful for situations where say we want to log information but don't want to display it to say the console. Ideally we should be able to control this on a severity level. For example, lets say you want to log information to trace that could be helpful to know in some situations but you don't want this information to be posted to the console. The ability to disable console logging for trace would be useful.
- [ ] **Add logging categories.** Logging categories can have multiple benefits for our logger. The simplest being that it can gives us pertinent information to where the log was actually sent from. Such as the rendering pipeline or the windowing system and can be used for even more advanced systems like log filtering in ImGUI.
- [ ] **Add asynchronous logging.** This is more of a nice to have but is not entirely required.

### PR DESCRIPTION
**Pull request by karnkaul**
Close #18 

Comments are highly encouraged, though keep in mind this is still WIP until marked ready for review.

Refer to linked issue and inline docs / comments for more context.

### CODE REVIEW

File: engine/include/logger/log.hpp
Comment (Rinzii) [inline engine/include/logger/log.hpp:9]: minor opinion on the matter, but I'd say we should instead use the monikers: trace, log, warn, error, and fatal.

Though personally I'm a bit indifferent to trace and log though I do find them more explicit in their purpose for logging. As for fatal, I feel this is a much more important log type to allow for the filtering of fatal crashes. Though it can be argued that error can also take up this torch. Though I'd like to also hear your opinion on the matter.

File: engine/include/logger/log.hpp
Comment (Rinzii) [inline engine/include/logger/log.hpp:12]: I want your thoughts on this, but even though allowing the user to type the category explicitly, I fear it could be bug-prone if a user fails to enter the consistent category. For example, what if one user puts [Rendering] and another puts [Graphics]?

This could possibly create issues with category filtering down the line. Trusting the developer always to do the right thing is generally not the best choice. :P

I think we should instead define log categories as an enum or something explicit that can be added too when we need more categories versus letting the user define it explicitly to avoid inconsistencies and to make our lives easier.

File: engine/src/logger/instance.cpp
Comment (Rinzii) [inline engine/src/logger/instance.cpp:30]: We have access to the date library through ext, which is the precursor to C++20's Chrono. All of the timezone and date functions there are fully thread-safe, and in instances like this, we should probably use that instead of dealing with the can of worms that std::gmtime and such open.

Here is the documentation to give you an idea of how it works as it's slightly different from how the standard implements it.

https://github.com/HowardHinnant/date/tree/master#readme

File: engine/src/logger/instance.cpp
Comment (Rinzii) [inline engine/src/logger/instance.cpp:54]: Is there any means to avoid this limitation or are we fully locked in by std::string?

**Review by Rinzii — CHANGES_REQUESTED** [review]
Overall everything looks pretty solid. Only a few comments I have for some sections of the code.

File: engine/include/logger/log.hpp
Comment (karnkaul) [reply engine/include/logger/log.hpp:9]: Fatal makes sense if it's different from error, which by your description it should be (by throwing an exception). Would asserts be discouraged in favour of `fatal` logs in that case? 

As for traces, I don't believe they are needed for basic client applications, as we can hook up a debugger / tracy / etc to the app at any time (unlike a remote app / server). It becomes useful when hard-to-reproduce Heisenbugs are involved, but if we get into that state we have much bigger problems to worry about anyway.

File: engine/include/logger/log.hpp
Comment (karnkaul) [reply engine/include/logger/log.hpp:12]: Adding `static constexpr std::string_view category_v{"Graphics"}` and enforcing logs in that domain to use it as the category argument solves that problem. Using enums does make it _impossible_ to pass in an invalid category, but at the cost of two things I think are worth keeping:
- Having to do zero maintenance on any other files when adding a new category to the codebase.
- Not having to prefix `logger::Category::eGraphics` (but instead `category_v`, with the enclosing namespace / class being implicit in that domain).

File: engine/src/logger/instance.cpp
Comment (karnkaul) [reply engine/src/logger/instance.cpp:30]: Gotcha, will take a look. I'd just gone off what exists in `log.cpp` for this.

File: engine/src/logger/instance.cpp
Comment (karnkaul) [reply engine/src/logger/instance.cpp:54]: Can't think of any means to avoid it except using a third party library for strings. We will need lots of fixed strings throughout UI code anyway, IMO having that class and building upon it later as needed is a better option for this app.

Comment (karnkaul) [reply engine/src/logger/instance.cpp:54]: [Context](https://github.com/ocornut/imgui/blob/master/docs/FAQ.md#q-how-can-i-interact-with-standard-c-types-such-as-stdstring-and-stdvector) for "lots of fixed strings throughout UI code":

> Dear ImGui applications often need to make intensive use of strings. It is expected that many of the strings you will pass to the API are raw literals (free in C/C++) or allocated in a manner that won't incur a large cost on your application. **Please bear in mind that using std::string on applications with a large amount of UI may incur unsatisfactory performances. Modern implementations of std::string often include small-string optimization (which is often a local buffer) but those are not configurable and not the same across implementations.**

File: engine/include/logger/log.hpp
Comment (karnkaul) [reply engine/include/logger/log.hpp:9]: PS: Will change it to `log()`, the reason I prefer `info()` is because `log()` is ambiguous with the maths interpretation. Also, `[E]`, `[W]`, `[I]` are recognized by many log readers / text editors and often coloured accordingly. Not sure about `[L]` and `[F]`.

Comment (karnkaul) [reply engine/include/logger/log.hpp:9]: Screenshot_20230716_011134
Data point (VSCode)

File: engine/src/logger/instance.cpp
Comment (karnkaul) [reply engine/src/logger/instance.cpp:30]: @Rinzii where's the date lib? I can't seem to find it in `ext`, either subdirs or CMake. 😄

File: engine/src/logger/instance.cpp
Comment (Rinzii) [reply engine/src/logger/instance.cpp:54]: I agree, a fixed string class will be super useful down the line and I do think we should keep it; my biggest concern is that a 57-character limit is quite limiting for a log message as shown below in the example. How hard would it be to implement a third-party string solution? If it would take longer than 1-2 hours I'd say we should say whatever ever and deal with it but I'd like to hear your thoughts if you think it'd be feasible to bring something in that can circumvent the formatting cutoff.

Also for reference, this is 55 characters counting spaces.
"I hope you are enjoying your time finding this helpful."

File: engine/include/logger/log.hpp
Comment (Rinzii) [reply engine/include/logger/log.hpp:12]: The string_view solution does seem like a valid option. @Kouros26 , what is your opinion?

File: engine/include/logger/log.hpp
Comment (Rinzii) [reply engine/include/logger/log.hpp:9]: Hmmm, it's rough for me as I want something that is explicit like fatal but I don't think allowing fatal itself to throw exceptions is the right choice. I'd say then you are right that if fatal is not different from error then it does not make sense. 

I'll say then we won't add a fatal as I don't like the idea of allowing fatal to actually perform exceptions.

As for trace, you have a valid point. I'll say then it's fine to exclude it.

File: engine/src/logger/instance.cpp
Comment (karnkaul) [reply engine/src/logger/instance.cpp:54]: The limit is for the format specification, not the log string, that is unlimited. 

PS: Just to clarify, the format spec limit is 64, while the _default_ format string, which includes all the "decorators", is 57 characters long. This is also static asserted in the header, so if we update the default format to cross that limit the compiler will tell us (and we simply double the capacity). If a format spec longer than 64 chars is set at runtime, the _spec_ will be cut off, which will result in some garbage in the logs, but the incoming log messages will still be printed in full.

Comment (karnkaul) [reply engine/src/logger/instance.cpp:54]:
```cpp
static constexpr std::string_view default_format_v{
  "[{level}][T{thread}] {message} [{timestamp}] [{location}]"
};
// ...
FixedString<format_size_v> format{default_format_v};
```
This part ^.

File: engine/src/logger/instance.cpp
Comment (Rinzii) [reply engine/src/logger/instance.cpp:30]: Ah, it was being added by FetchContent according to the suggestions from the lib owner, but it seems it got lost in the move to the new repo. Sorry about that @karnkaul !

Just add this below to ext and the link is date::date

`
     include( FetchContent )
     FetchContent_Declare( date_src
       GIT_REPOSITORY https://github.com/HowardHinnant/date.git
       GIT_TAG        v3.0.1  # adjust tag/branch/commit as needed
     )
     FetchContent_MakeAvailable(date_src)
     ...
     target_link_libraries (my_target PRIVATE date::date)
`

File: engine/src/logger/instance.cpp
Comment (Rinzii) [reply engine/src/logger/instance.cpp:54]: Aaaaaah, ok, that clears things up much more. Sorry for my misunderstanding! I have no issue with the code then.

File: engine/src/logger/instance.cpp
Comment (karnkaul) [reply engine/src/logger/instance.cpp:30]: No problem! However, if it's not part of the project yet, I'm gonna argue that it's out of scope of this PR. Happy to create a new Issue tracking that enhancement, and covering in subsequent PRs.

File: engine/include/logger/log.hpp
Comment (Kouros26) [reply engine/include/logger/log.hpp:12]: I also prefer the string_view option now that I've read karnage's explanation

File: engine/include/logger/instance.hpp
Comment (Kouros26) [inline engine/include/logger/instance.hpp:47]: setConfig and addSink can be made const

File: engine/src/logger/instance.cpp
Comment (Rinzii) [reply engine/src/logger/instance.cpp:30]: That would be acceptable ^^

File: engine/include/logger/instance.hpp
Comment (karnkaul) [reply engine/include/logger/instance.hpp:47]: I don't think we want users to be able to modify the state of `const` instances. The fact that it uses PIMPL is just an implementation detail, changes in state to `Impl` that have public side effects should require a mutable `Instance`.

**Review by Rinzii — DISMISSED** [review]

**Comment by karnkaul** [comment]
Sorry @Rinzii , need another review after resolving the merge conflict 😛

**Comment by karnkaul** [comment]
@Rinzii @Kouros26 Let's get this in before it becomes out of date again?

**Review by Kouros26 — APPROVED** [review]

**Pull request merged**
